plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/KuduRecordCursor.java

```diff
@@ -153,12 +157,20 @@ private PartialRow buildPrimaryKey()
     @Override
     public void close()
     {
-        try {
-            scanner.close();
+        if (!closed) {
+            try {
+                scanner.close();
+            }
+            catch (KuduException e) {
+                throw new RuntimeException(e);
+            }
+            finally {
+                currentRow = null;
+                closed = true;
+            }
         }
-        catch (KuduException e) {
-            throw new RuntimeException(e);
+        else {
+            throw new TrinoException(GENERIC_INTERNAL_ERROR, "Cursor already closed");
```

Comment:
throw vs log all previous `close` calls (i.e. keep track of all callers and log them if close is called 2nd or more times)?

Asking because the throw stack trace won't tell us which "other" thread closed the cursor - it'll only tell us that we closed it twice and who closed it the 2nd time.

I'm assuming this is for a debug build we can use to find the actual reason why this happens.

Reply:
yeah, I think just logging it (and the previous stack-trace of who called) should be sufficient to find why it's happening.

So maybe we can log + not fail the query until we find the cause.

Reply:
But from Java docs for KuduScanner#close it says 
```
   * Closes this scanner (don't forget to call this when you're done with it!).
   * <p>
   * Closing a scanner already closed has no effect.
```

Reply:
As far as I see their close method looks like this : 
```
public Deferred<RowResultIterator> close() {
  if (closed) {
    return Deferred.fromResult(null);
  }
  return client.closeScanner(this).addCallback(closedCallback()); // TODO errBack ?
}
```

So second close is guarded by variable close.

But looks like this variable is not thread safe:

`private boolean closed = false;`

Reply:
The Scanner itself isn't thread-safe - the class level javadoc says so. So maybe we are sharing it across threads in some way leading to this double-close.

Reply:
Ok so this looks like valid explanation.
- we read from one thread, reach end of keys scanner is closed automatically 
- then we do first time close of RecordPageSource, which close already closed scanner but from other thread, just because it's not guarded  internally we fail

Reply:
We have 1.15.0 and it contains the same code that you pointed, thanks !

Reply:
But we don't use that `AysncKuduScanner` directly rather we use `KuduScanner` which claims it provides a `Synchronous version of {@link AsyncKuduScanner}. Offers the same API but with blocking methods.` - but I'm not sure if it provides a synchronous version.

Reply:
As far as I see AsyncKuduScanner (which is not thread safe by documentation) returns all operation as Deferred, as far as I see it's kind of analogous of Future.

KuduScanner scanner wraps almost all methods of AsyncKuduScanner and do join on Deferred's from AsyncKuduScanner. (looks like it's analogous of Future.get())

However I am not 100% sure, that synchronisation guarantees are as strong as Future.get.

First of all, it has some strange description:
```
Contract and Invariants

- Only one thread at a time is going to execute the callback chain.
- Each action taken by a callback happens-before  the next callback is invoked. In other words, if a callback chain manipulates a variable (and no one else manipulates it), no synchronization is required  
```

I dived a little bit in the code but cannot guarantee that I understood it fully, but noticed some strange things:
- like executing method itself is under synchronised block and adding callbacks to callback list also, but executing callback itself is not (but for example we are setting closed flag inside callback not inside method itself)

So as bottom line I am not sure that closed variable in AsyncKuduScanner via KuduScanner is guarded properly.
And I think this PR is wort to try.

Reply:
```
KuduScanner

public RowResultIterator close() throws KuduException {
    return KuduClient.joinAndHandleException(asyncScanner.close());
  }
```
and this code call next one:

```
AsyncKuduScanner

public Deferred<RowResultIterator> close() {
    if (closed) {
      return Deferred.fromResult(null);
    }
    return client.closeScanner(this).addCallback(closedCallback()); // TODO errBack ?
  }
```

So as we see here access to variable closed is not guarded in anyway